File: test/ModularPipelines.UnitTests/DependencyInjectionTests.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using ModularPipelines.DependencyInjection;
using ModularPipelines.TestHelpers;
using Moq;
using IConfiguration = Microsoft.Extensions.Configuration.IConfiguration;

namespace ModularPipelines.UnitTests;

public class DependencyInjectionTests
{
    [Test]
    public async Task AllDependenciesCanBeBuilt()
    {
        var host = await TestPipelineHostBuilder.Create().BuildHostAsync();

        var services = host.Services;

        var collection = services.GetRequiredService<IPipelineServiceContainerWrapper>().ServiceCollection;

        foreach (var serviceDescriptor in collection.Where(sd => sd.ServiceType.Namespace?.StartsWith("ModularPipeline") == true && !sd.ServiceType.IsGenericType))
        {
            services.GetRequiredService(serviceDescriptor.ServiceType);
        }
    }

    [Test]
    public void Validate()
    {
        var serviceCollection = new ServiceCollection()
            .AddSingleton(Mock.Of<IHost>())
            .AddSingleton(Mock.Of<IHostEnvironment>())
            .AddSingleton(Mock.Of<IConfiguration>());

        DependencyInjectionSetup.Initialize(serviceCollection);

        serviceCollection.BuildServiceProvider(new ServiceProviderOptions
        {
            ValidateScopes = true,
            ValidateOnBuild = true,
        });
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Engine/DependencyInjectionTests.fs
namespace ModularPipelines.UnitTests.FSharp.Engine

open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open ModularPipelines.Context
open ModularPipelines.Extensions
open ModularPipelines.DependencyInjection
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open Moq
open TUnit.Assertions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core
open TUnit.Assertions.Extensions

type private DependencyInjectionModule() =
    inherit Module<bool>()

    override _.ExecuteAsync(_: IModuleContext, _) =
        System.Threading.Tasks.Task.FromResult(true)

type DependencyInjectionTests() =
    [<Test>]
    member _.AllDependenciesCanBeBuilt() = async {
        let! host =
            TestPipelineHostBuilder.Create()
                .AddModule<DependencyInjectionModule>()
                .BuildHostAsync()
            |> Async.AwaitTask

        let services = host.Services
        let collection = services.GetRequiredService<IPipelineServiceContainerWrapper>().ServiceCollection

        for serviceDescriptor in
            collection
            |> Seq.filter (fun sd ->
                let serviceNamespace = sd.ServiceType.Namespace
                not sd.ServiceType.IsGenericType
                && not (isNull serviceNamespace)
                && serviceNamespace.StartsWith("ModularPipeline")) do
            services.GetRequiredService(serviceDescriptor.ServiceType) |> ignore

        do! check(Assert.That(true).IsTrue())
    }

    [<Test>]
    member _.Validate() = async {
        let serviceCollection =
            ServiceCollection()
                .AddSingleton(Mock.Of<IHost>())
                .AddSingleton(Mock.Of<IHostEnvironment>())
                .AddSingleton(Mock.Of<IConfiguration>())

        DependencyInjectionSetup.Initialize(serviceCollection)

        serviceCollection.BuildServiceProvider(
            ServiceProviderOptions(
                ValidateScopes = true,
                ValidateOnBuild = true
            )
        )
        |> ignore

        do! check(Assert.That(true).IsTrue())
    }
